### ISSUE TITLE
Update docs for Next.js

### ISSUE THREAD

**Issue by SlavenDj** [supabase/supabase#27666]
# Improve documentation

## Link 

Docs from the homepage inside dashboad for connecting Next.js with Supabase
![image](https://github.com/supabase/supabase/assets/53652906/81163cc4-e00e-4843-ac77-f917d1d967d6)

## Describe the problem

Gives a warning about declaration of some methods
![image](https://github.com/supabase/supabase/assets/53652906/44eaf00f-8639-4056-a2f1-021a60a746ae)

## Describe the improvement

Update the Next.js code or/and add option to swich between verstions of package

**Comment by josephmarkus** [comment]
@SlavenDj docs are up-to-date on the auth page for Next.js: https://supabase.com/docs/guides/auth/server-side/nextjs

You could raise a PR for this 💯 

<img width="1334" alt="Screenshot 2024-07-01 at 11 08 36" src="https://github.com/supabase/supabase/assets/9258365/afef76f6-a2c9-4379-8493-acb699dee8bf">

**Comment by SlavenDj** [comment]
@josephmarkus Thanks for the info, but I still got an error from that code:
![image](https://github.com/supabase/supabase/assets/53652906/dc94c7be-e4f7-4bab-9988-aa65ab38ae09)
then createClient has to be async and that is weird, I hope I am doing something wrong

**Comment by ErikPetersenDev** [comment]
@SlavenDj  Can you share how to get the error above to show up?  I've been changing some things in my app to get it to trigger but haven't been able to replicate that error yet.  I checked and my `createClient` function in `server.ts` is not async.

Just an FYI, I have an open PR also related to Next.js docs.  I've been trying to get this all working as well (after updating to `@supabase/ssr` v0.4.0). https://github.com/supabase/supabase/pull/27659.   Mine seems to be working at the moment but I may not have tested every combination.

One thing that I'm noticing in https://supabase.com/docs/guides/auth/server-side/nextjs is that it seems like there are two patterns for protecting routes and redirecting, both within the same document.  The middleware checks for a user and redirects to `/login` if unauthenticated.  This is the area of code my PR above (and one from Friday) was looking to get working.  This focuses on using middleware as the arbiter of which pages are protected or not.

Then lower in the same doc there's an example of checking for a user and redirecting to login from within the page component itself.

It's not clear to me which is the preferred method, but if there is a check in middleware then it seems like the extra code within the page would be redundant (extra trip to the auth server as well).  I've been using the Clerk docs as a point of comparison and it appears they use middleware exclusively to protect routes in Next.js (they have a quickstart [here](https://clerk.com/docs/quickstarts/nextjs) and docs on protecting routes with middleware [here](https://clerk.com/docs/references/nextjs/clerk-middleware)).

**Comment by SlavenDj** [comment]
@ErikPetersenDev I have just pasted that code and when i start a server that is the first error. Like I go to the login page create a client and 💥, error... I am using old style of the creating client for now because i have not idea what is going on. I had to put it async because it says it should be server action. Do you need anything more specific?

**Comment by ErikPetersenDev** [comment]
Interesting, thanks!  That definitely helps.  

One issue at the moment I think is that parts of the docs are overlapping across multiple pages, plus the `with-supabase` template, and also people are using various versions (before/after 0.4.0) so it's hard to track exactly what the best/recommended solution is at the moment.  I might try to start a new next.js app without the template and follow the example to see if I can find additional problems (I used the with-supabase template and made changes to get it working, then updated when 0.4.0 came out).  I'm not actively troubleshooting at the moment though (also not a member of SB team if that wasn't clear), but sometime later today I can see what I find.

**Comment by charislam** [comment]
Thanks for the explanations and for your PRs, @ErikPetersenDev! We're working on standardizing the instructions across all the places where they've available (we realized we're missing a few). The [Server-Side Auth for Next.js guide](https://supabase.com/docs/guides/auth/server-side/nextjs?queryGroups=router&router=app) should contain the most recent version, for anyone who's confused about which to follow while we clean things up!

**Comment by wadjeroudi** [comment]
same issue using the new middleware
![image](https://github.com/supabase/supabase/assets/7465362/65160c1f-53a6-4ce7-b70d-7437a8d2e965)

```js
export async function getProjectById(id: string, type: string): Promise<Project | null> {
  const supabase = createClient();

  console.log('get ID', id, type)
  console.log('get USER', await supabase.auth.getUser());
```

the getUser on server side seems to fire randomly a setAll call, to set a cookie in vercel env

```js
tD [Error]: Cookies can only be modified in a Server Action or Route Handler. Read more: https://nextjs.org/docs/app/api-reference/functions/cookies#cookiessetname-value-options
    at Proxy.callable (/var/task/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:13:13001)
    at /var/task/.next/server/app/[locale]/rtp/[id]/page.js:1:6846
    at Array.forEach (<anonymous>)
    at setAll (/var/task/.next/server/app/[locale]/rtp/[id]/page.js:1:6806)
    at g (/var/task/.next/server/chunks/721.js:1:3145)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Object.callback (/var/task/.next/server/chunks/721.js:5:4926)
    at async /var/task/.next/server/chunks/814.js:1:107088
    at async Promise.all (index 1)
    at async a_._notifyAllSubscribers (/var/task/.next/server/chunks/814.js:1:107135)

**Comment by ErikPetersenDev** [comment]
A couple other questions I thought of:

Does this only happen in Vercel or does it happen when running locally too?

By any chance is your server component that's calling `getUser` in a layout?  In this Discussion a couple others working through this same problem said that was the case for them: https://github.com/orgs/supabase/discussions/27037#discussioncomment-9914146.

**Comment by wadjeroudi** [comment]
no problem locally, it happens on vercel because of their serverless function and limitation.
calling from a server component (getUser is called inside getProjectById)

```js
export default async function ProjectPage({params}: any) {
  const project = await getProjectById(params.id as string, ProjectTypeRTP);
```

the new middleware
```js
import { NextRequest } from 'next/server';
import { createI18nMiddleware } from 'next-international/middleware';

import { createServerClient } from '@supabase/ssr';

const handleI18nRouting = createI18nMiddleware({
  locales: ['en', 'fr'],
  defaultLocale: 'en',
  urlMappingStrategy: 'rewriteDefault',
});

export async function middleware(request: NextRequest) {
  const response = handleI18nRouting(request);

  const supabase = createServerClient(
    process.env.NEXT_PUBLIC_SUPABASE_URL!,
    process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!,
    {
      cookies: {
        getAll() {
          return request.cookies.getAll()
        },
        setAll(cookiesToSet) {
          cookiesToSet.forEach(({ name, value, options }) => request.cookies.set(name, value))
          cookiesToSet.forEach(({ name, value, options }) =>
            response.cookies.set(name, value, options)
          )
        },
      },
    }
  )

  const {data: {user}} = await supabase.auth.getUser();

  return response;
}

export const config = {
  matcher: [
    '/',
    '/((?!_next/static|_next/image|favicon.ico|.*\\.(?:svg|png|jpg|jpeg|gif|webp)$).*)',
  ],
};
```

thx.

**Comment by ErikPetersenDev** [comment]
@wadjeroudi Here are some thoughts:

My understanding is that the middleware call to `getUser` is to ensure that if any changes are needed to cookies, they are set in middleware before the request makes it to any server components.  By doing that, when a `getUser` call is made from a server component, it should theoretically never result in a situation that would need to set cookies, since any token refreshes, etc. would have already happened in middleware.

Therefore, one possibility could be that once in a while an error occurs in the middleware `getUser` call but we continue on to the server components anyway.  In that case, if we're in a situation where cookie updates are required (token refresh, etc), the server component would then try to set them and throw the exception you're seeing.

One thing you could try is checking for an `error` in your `getUser` response within middleware and either logging it or maybe even failing the request at that point.  Looking at the getUser code, an "AuthError" will come back in the response, but a different error would be thrown (although I believe that'd already be unhandled and fail the request in middleware, so that likely isn't happening).  If you do that you could potentially rule this out as the cause.

Another option would be to create a separate server client for use in server components, which leaves off the "setAll" definition (allowed by the signature).  That won't really tell us anything new though because the client will log a console warning that it needed to set cookies and couldn't.  You may get odd authentication behavior in that case though... if server components are trying to set cookies you _want_ to fail in my opinion.

Here's a rough example of the type of handling that you could add to your middleware to maybe better see if you're hitting `getUser` issues periodically.  If it _does_ turn out to be transient getUser errors in middleware then a retry strategy might be required.

```
    const { error } = await supabase.auth.getUser();
    if (error) {
      console.log(
        "Auth error retrieving user in middleware: ",
        JSON.stringify(error)
      );

      // Optionally fail the request.  If you don't and the `getUser` call fails in a situation
      // where cookies need to be set, you could end up hitting an error when the request
      // gets to a server component that calls `getUser`. However, if transient failures
     // are happening frequently then this could cause unnecessary failed requests.
      return NextResponse.json(
        { message: "Internal server error" },
        { status: 500 }
      );
    }
```

**Comment by ErikPetersenDev** [comment]
@wadjeroudi Based on your code I did some testing and I was able to replicate the issue.

TL;DR:  Your `setAll` function definition needs to be updated in one spot to ensure that a refreshed cookie is sent to the server.

Current:
```
        setAll(cookiesToSet) {
          cookiesToSet.forEach(({ name, value, options }) => request.cookies.set(name, value))
          cookiesToSet.forEach(({ name, value, options }) =>
            response.cookies.set(name, value, options)
          )
        },
```
What's needed:
```
        setAll(cookiesToSet) {
          cookiesToSet.forEach(({ name, value, options }) => request.cookies.set(name, value))
          response = handleI18nRouting(request);  /// <---- Add this line
          cookiesToSet.forEach(({ name, value, options }) =>
            response.cookies.set(name, value, options)
          )
        },
```

I haven't tested this exact fix (with the extra I18n middleware), but it should work.  The docs show `supabaseResponse = NextResponse.next({ request });` on that line of code, which is what I was using.  I think the key thing is that you need to recreate your new response after updating the request cookies, which in your case would mean re-running the middleware.

I did a lot of troubleshooting and testing, shortening my token lifetime to 60 seconds and adding logging.  If you leave that line out and a token refresh occurs, the response cookies will be correctly updated, but the request cookie that is sent on to the server will still be the old token.  Because of this, the server component tries to refresh, gets the new token (since refresh tokens can be used more than once within 10 seconds), and then tries to set it on the server (crashing since it can't).  This results in a 500, but the updated cookie is _still set in the response_.  Therefore, going to the same page immediately after will work normally since the correct/refreshed token is in the second request.
